storage/innobase/xtrabackup/src/xtrabackup.cc

```diff
@@ -5486,59 +5486,141 @@ static std::string read_file_as_string(const std::string file) {
   fclose(f);
   return std::string(content, len);
 }
-/* Handle DDL for renamed files */
+
+/**
+Handle DDL for renamed files
+example input: test/10.ibd.ren file with content = test/new_name.ibd ;
+      -> tablespace with space_id = 10 will be renamed to test/new_name.ibd
+@return true on success */
 static bool prepare_handle_ren_files(
     const datadir_entry_t &entry, /*!<in: datadir entry */
-    void *arg __attribute__((unused))) {
+    void * /*data*/) {
   if (entry.is_empty_dir) return true;
 
-  std::string ren_file = entry.path;
-  std::string ren_path = entry.rel_path;
-  std::string from_base = entry.datadir;
-  std::string to_base = entry.datadir;
-#ifdef UNIV_DEBUG
-  size_t index = ren_file.find(".ren");
-  assert(index != std::string::npos);
-#endif  // UNIV_DEBUG
-  std::string from = ren_path.substr(0, ren_path.length() - 4);
-  from_base += from;
-  std::string to = read_file_as_string(ren_file);
-  to_base += to;
-  if (to.empty()) {
-    xb::error() << "Can not read " << ren_file;
+  char dest_space_name[FN_REFLEN];
+  space_id_t source_space_id;
+  std::string ren_file_content;
+  std::string ren_file_name = entry.file_name;
+  std::string ren_path = entry.path;
+
+  Fil_path::normalize(ren_path);
+  // trim .ibd.ren
+  source_space_id =
+      atoi(ren_file_name.substr(0, ren_file_name.length() - 8).c_str());
+  ren_file_content = read_file_as_string(ren_path);
+
+  if (ren_file_content.empty()) {
+    xb::error() << "prepare_handle_ren_files: " << ren_path << " is empty.";
     return false;
   }
-  xb::info() << "prepare_handle_ren_files: From: " << from << " To: " << to;
-  rename_force(from, to);
-  if (xtrabackup_incremental) {
-    rename_force(from_base + ".delta", to_base + ".delta");
-    rename_force(from_base + ".meta", to_base + ".meta");
+  // trim .ibd
+  snprintf(dest_space_name, FN_REFLEN, "%s",
+           ren_file_content.substr(0, ren_file_content.length() - 4).c_str());
+
+  fil_space_t *fil_space = fil_space_get(source_space_id);
+
+  if (fil_space != NULL) {
+    char tmpname[FN_REFLEN];
+    char *oldpath, *space_name;
+    bool res =
+        fil_space_read_name_and_filepath(fil_space->id, &space_name, &oldpath);
+
+    if (!res || !os_file_exists(oldpath)) {
+      xb::error() << "prepare_handle_ren_files: Tablespace " << fil_space->name
+                  << " not found.";
+      ut::free(oldpath);
+      ut::free(space_name);
+      return false;
+    }
+
+    strncpy(tmpname, dest_space_name, sizeof(tmpname) - 1);
```

Comment:
why are we using this tmpname variabel? can we directly pass dest_space_name to fil_rename_tablespace.